src/Microsoft.DotNet.Build.Tasks/PackageFiles/codeOptimization.targets

```diff
@@ -0,0 +1,81 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <UsingTask TaskName="ExtractZipFile" AssemblyFile="$(BuildToolsTaskDir)Microsoft.DotNet.Build.Tasks.dll"/>
+
+  <PropertyGroup>
+    <OptimizationDataVersion>1.0.0-prerelease</OptimizationDataVersion>
+    <OptimizationDataPackageName>Microsoft.DotNet.BuildTools.OptimizationData</OptimizationDataPackageName>
+    <OptimizationDataDir>$(ToolsDir)/OptimizationData</OptimizationDataDir>
+  </PropertyGroup>
+
+  <!-- We should only run this target on Windows and only if EnableProfileGuidedOptimization is set and we have training data -->
+  <Target Name="OptimizeWithTrainingData" AfterTargets="AfterBuild" BeforeTargets="CopyFilesToOutputDirectory" DependsOnTargets="RestoreOptimizationDataPackage"
+          Condition="'$(OSEnvironment)'=='Windows_NT' and '$(EnableProfileGuidedOptimization)'=='true' and Exists('$(OptimizationDataDir)/$(AssemblyName).zip')">
+
+    <PropertyGroup>
+      <IBCMergeToolPath>$(ToolsDir)/ibcmerge.exe</IBCMergeToolPath>
+      <OptimizationDataForAssemblyDir>$(IntermediateOutputPath)/OptimizationData/$(AssemblyName)</OptimizationDataForAssemblyDir>
+      <OptimizedAssemblyFile>$(IntermediateOutputPath)/OptimizedAssembly/$(AssemblyName).dll</OptimizedAssemblyFile>
+    </PropertyGroup>
+
+    <!-- Extract the profile data from the archive -->
+    <ExtractZipFile SourceArchive="$(OptimizationDataDir)/$(AssemblyName).zip"
+                    DestinationDirectory="$(OptimizationDataForAssemblyDir)"
+                    OverwriteDestination="true" />
+
+    <!-- Copy the compiled assembly into a folder for further processing -->
+    <MakeDir Directories="$(IntermediateOutputPath)/OptimizedAssembly" />
+    <Message Text="Copy @(IntermediateAssembly) to $(IntermediateOutputPath)/OptimizedAssembly" />
```

Comment:
perhaps we can set this message important to low so that it is only present on diagnostic builds